### PR TITLE
DeepChem DFT

### PR DESCRIPTION
**Pull request by sudo-rsingh**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by sudo-rsingh** [comment]
1. .github/workflows/dqc_install.yml

  - Expanded Python version testing: Added Python 3.9, 3.10, and 3.11 to the test
  matrix (previously only 3.8)
  - Added system dependencies: Installed cmake, g++, libopenblas-dev, and
  liblapack-dev via apt-get
  - Removed DQC dependency checks: Removed verification steps for dqc and xitorch
  imports, keeping only pylibxc check

  2. deepchem/feat/dft_data.py

  - Refactored imports: Replaced direct dqc imports with internal
  deepchem.utils.dft_utils imports for Mol, BaseSystem, BaseGrid
  - Return type changes: Updated get_val() methods to return torch.Tensor instead
  of np.ndarray for better gradient tracking
  - Removed numpy conversions: Changed from np.array([val.tolist()]) to native
  torch tensor operations with .unsqueeze(0)
  - Maintained gradient flow: Ensured tensors retain computational graphs for
  backpropagation

  3. deepchem/feat/tests/test_dftdata.py

  - Updated imports: Changed from dqc.qccalc.ks to deepchem.utils.dft_utils.KS
  - Better error handling: Added numpy import and improved exception handling with
  warnings
  - Renamed flag: Changed has_dqc to has_torch for more accurate dependency
  tracking

  4. deepchem/feat/tests/test_dftdata2.py

  - Fixed tensor comparison: Added .detach().numpy() call before numpy allclose
  comparison to handle torch tensors properly

  5. deepchem/models/dft/dftxc.py

  - Simplified tensor handling: Removed unnecessary torch.as_tensor() and
  torch.tensor() conversions since get_val() now returns torch tensors
  - Preserved gradients: Removed explicit requires_grad=True as gradients are
  automatically tracked in the new implementation

  6. deepchem/models/dft/nnxc.py

  - Refactored imports: Consolidated imports from internal utils instead of
  external dqc library
  - Added family property: Implemented family property for both NNLDA (returns 1)
  and NNPBE (returns 2) to identify XC functional types
  - Type annotations: Added explicit type hints for nnxc attribute as Union[NNLDA, 
  NNPBE]

  7. deepchem/models/dft/scf.py

  - Import consolidation: Moved from external dqc imports to internal
  deepchem.utils.dft_utils imports
  - Removed warnings: Cleaned up try-except blocks as dependencies are now internal
  - Type hint fix: Updated get_xc() return type to Union[BaseNNXC, HybridXC] from
  just HybridXC

  8. deepchem/models/tests/test_nnxc.py

  - Import path updates: Changed ValGrad import from dqc.utils.datastruct to
  deepchem.utils.dft_utils

  9. deepchem/utils/cache_utils.py

  - Documentation fix: Updated docstring examples to use correct function name
  normalize_prefix instead of _normalize_prefix

  10. deepchem/utils/dft_utils/api/parser.py

  - Integer atomic number handling: Modified parse_moldesc() to keep atomic numbers
   as integers when appropriate, avoiding unnecessary float conversion
  - Fractional mode fix: Only converts to dtype for explicit fractional atomic
  numbers, preventing unintended fraction mode activation
  - Enhanced documentation: Added more detailed docstring with better formatting
  and examples
  - Output consistency: Ensures atomzs tensor is integer type for whole atomic
  numbers

  11. deepchem/utils/dft_utils/hamilton/hcgto.py

  - Density matrix unconversion: Added dm = self._orthozer.unconvert_dm(dm) in
  aodm2dens() method to properly transform density matrix before computing electron
   density

  12. deepchem/utils/dft_utils/qccalc/scf_qccalc.py

  - Added getparamnames() method: Implements parameter name retrieval for given
  methods with optional prefix support, delegating to the engine's implementation

  13. deepchem/utils/dft_utils/system/mol.py

  - Added requires_grid() method: Returns True to indicate the molecular system
  requires a grid for calculations
  - Added make_copy() method: Implements system cloning with parameter overrides,
  allowing flexible system duplication with modified attributes
  - Fixed docstring example: Corrected expected output from [1.5000, 0.5000] to
  [2.] for orbital weights

  14. deepchem/utils/dft_utils/xc/libxc_wrapper.py

  - Tensor conversion fix: Changed res[0] * rho to res[0] * torch.as_tensor(rho) to
   ensure consistent tensor types in operations

  15. deepchem/utils/dftutils.py

  - Type checking imports: Added TYPE_CHECKING for proper type hint imports without
   circular dependencies
  - Improved type hints: Updated KSCalc.__init__() to accept both BaseQCCalc types
  - Added type assertions: Included explicit isinstance check for mypy to
  understand tensor types in aodmtot()
  - Type annotation for dmtot: Added explicit type hint for dmtot variable

  16. deepchem/utils/safeops_utils.py

  - Relaxed type constraint: Changed safepow() parameter p from torch.Tensor to
  Union[float, torch.Tensor] to accept scalar powers
  - Updated documentation: Modified docstring to reflect the parameter can now be a
   float or tensor

  17. deepchem/utils/test/test_dft_utils.py

  - Updated expected values: Changed expected atomzs from float tensors [1., 9.] to
   integer tensors [1, 9]
  - Fixed orbital weight test: Updated expected orbital weights from [1.5000, 
  0.5000] to [2.0] and removed spin=1 parameter

  18. requirements/env_dqc.yml

  - Replaced external dependencies: Removed dqc, xitorch, torch-geometric, and
  pylibxc2
  - Added internal dependencies: Added dqclibs (from GitHub) and libxc (from
  GitLab) as direct installations
  - Added utilities: Added tqdm for progress bars
  - Kept pyscf: Retained pyscf[all] for quantum chemistry calculations

**Comment by JoseAntonioSiguenza** [comment]
Test for 3,9 and 3,11 documents are failing `Test for documents / docs-build (3.9) (pull_request)`. This might not be related to these changes, as it appears in other PRs, but is a failing test.

**Comment by JoseAntonioSiguenza** [comment]
`Unit Tests / unit-tests (ubuntu-latest, 3.8) (pull_request)` is breaking when installing dependencies.

**Comment by JoseAntonioSiguenza** [comment]
There is also a new failing test. Bash build is failing to install `torch-cluster` for macOS.

**Comment by JoseAntonioSiguenza** [comment]
These are great changes! This PR is very complex to review since there are various parts being modified, but I think most of the code is working. I'd suggest splitting this into 3 PRs (1. env and install changes, 2. deepchem utils, 3. deepchem models and featurizers)

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py [hunk 1]
```diff
@@ -313,6 +313,24 @@ def dm2energy(self, dm: Union[torch.Tensor, SpinParam[torch.Tensor]]):
             (isinstance(dm, SpinParam) and self._polarized), type(dm)
         return self._engine.dm2energy(dm)
 
+    def getparamnames(self, methodname: str, prefix: str = "") -> List[str]:
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/dft_utils/qccalc/scf_qccalc.py:316]: Can you please change this function name to match other deepchem function name styles? Maybe "get_param_names"?

Comment (JoseAntonioSiguenza) [reply deepchem/utils/dft_utils/qccalc/scf_qccalc.py:316]: Same for `methodname` parameter.

File: requirements/env_dqc.yml [hunk 2]
```diff
@@ -9,11 +9,10 @@ dependencies:
...
-   - torch-geometric
-   - git+https://github.com/diffqc/dqc.git
-   - xitorch
-   - pylibxc2
+   - git+https://github.com/diffqc/dqclibs.git
+   - git+https://gitlab.com/libxc/libxc.git
    - pytest
    - PyYAML
    - yamlloader
+   - tqdm
```
Comment (JoseAntonioSiguenza) [inline requirements/env_dqc.yml:17]: Can you please share why adding `tqdm`? I believe this library is not being used.

**Comment by JoseAntonioSiguenza** [comment]
> * Test for documents / docs-build (3.9) (pull_request)

Actually, I believe these failures are related to changes in this PR since the error is no space left in device.

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py (hunk 1 above)
Comment (sudo-rsingh) [reply deepchem/utils/dft_utils/qccalc/scf_qccalc.py:316]: these were adopted from the dqc standards, now there are more than 20-30 instances where they were used in dft_utils directory.

File: requirements/env_dqc.yml (hunk 2 above)
Comment (sudo-rsingh) [reply requirements/env_dqc.yml:17]: `tqdm` is actually a mandatory requirement of deepchem, we cannot run deepchem without it (pls try in a new conda env to understand what i am saying here).
Earlier it was hidden in background by torch-geometric, so when i removed it from env-dqc, i needed to add tqdm.

**Comment by sudo-rsingh** [comment]
> > * Test for documents / docs-build (3.9) (pull_request)
> 
> Actually, I believe these failures are related to changes in this PR since the error is no space left in device.

I think that was a random github action issue, it didn't happen in next commit.

**Comment by sudo-rsingh** [comment]
@shreyasvinaya I wanted your opinion on the macos build script ci [1].
The last PR which triggered this was yours [2] and you know better about CI than me.
Should we try to fix this error or is it fine to merge with it as mac os is known to be unstable for deepchem.

[1].. https://github.com/deepchem/deepchem/actions/runs/19639878612/job/56240180534?pr=45
[2].. https://github.com/deepchem/deepchem/pull/4553

**Comment by sudo-rsingh** [comment]
Continued to #4601
